reco_utils/dataset/covid_utils.py

```diff
@@ -0,0 +1,180 @@
+        cols (list of str): Name of columns in which to look for duplicates.
+    
+    Returns:
+        df (pd.DataFrame): Pandas dataframe with duplicate rows dropped.
+    
+    """
+    for col in cols:
+        # Reset index
+        df = df.reset_index(drop=True)
+        
+        # Find where the identifier variable is duplicated
+        dup_rows = np.where(df.duplicated([col])==True)[0]
+        
+        # Drop duplicated rows
+        df = df.drop(dup_rows)
+
+    return df
+
+def remove_nan(df, cols):
+    """ Remove rows with NaN values in specified column.
+    
+    Args:
+        df (pd.DataFrame): Pandas dataframe.
+        cols (list of str): Name of columns in which to look for NaN.
+    
+    Returns:
+        df (pd.DataFrame): Pandas dataframe with invalid rows dropped.
+    
+    """
+    for col in cols:
+        # Convert any empty string cells to nan
+        df[col].replace('', np.nan, inplace=True)
+        
+        # Remove NaN rows
+        df = df[df[col].notna()]
+
+    return df
+
+def clean_dataframe(df):
+    """ Clean up the dataframe.
+    
+    Args:
+        df (pd.DataFrame): Pandas dataframe.
+    
+    Returns:
+        df (pd.DataFrame): Cleaned pandas dataframe.
+    """
+
+    # Remove duplicated rows
+    cols=['cord_uid','doi']
+    df = remove_duplicates(df, cols)
+    
+    # Remove rows without values in specified columns
+    cols=['cord_uid','doi','title','license','url']
+    df = remove_nan(df, cols)
+    
+    return df
+
+def retrieve_text(entry, blob_service, container_name):
+    """ Retrieve body text from article of interest.
+    
+    Args:
+        entry (pd.Series): A single row from the dataframe (df.iloc[n]).
+        blob_service (azure.storage.blob.BlockBlobService): Azure BlockBlobService for dataset.
+        container_name (str): Azure storage container name.
+    
+    Results:
+        text (str): Full text of the blob as a single string.
+    """
+    
+    try:
+        # select based on whether it's pdf or pmc_xml
+        if entry['has_pdf_parse'] == True:
+            blob_name = '{0}/pdf_json/{1}.json'.format(entry['full_text_file'], entry['sha'])
+        else:
+            if entry['has_pmc_xml_parse'] == True:
+                blob_name = '{0}/pmc_json/{1}.xml.json'.format(entry['full_text_file'], entry['pmcid']) 
+            else:
+                print('Neither PDF or PMC_XML data is available for this file')
+
+        # Extract text
+        data = json.loads(blob_service.get_blob_to_text(container_name=container_name, blob_name=blob_name).content)
+    
+        # the text itself lives under 'body_text'
+        full_text = data['body_text']
+
+        # many NLP tasks play nicely with a list of sentences
+        sentences = list()
+        for paragraph in full_text:
+            sentences.append(paragraph['text'])
+            
+        text = ''
+        for sentence in sentences:
+            text = text + ' ' + sentence
```

Comment:
there was an earlier comment on simplifying this with list comprehension, which is _usually_ faster. https://www.linkedin.com/pulse/list-comprehension-python-always-faster-than-alex-falkovskiy/

did that not work for some reason?
```suggestion
        # the text itself lives under 'body_text'
        full_text = data['body_text']

        # many NLP tasks play nicely with a list of sentences
        sentences = list()
        for paragraph in full_text:
            sentences.append(paragraph['text'])
            
        text = ''
        for sentence in sentences:
            text = text + ' ' + sentence
```
text = " ".join([sentence for sentence in data["body_text"])

Reply:
Sorry about missing this! The suggested change does not work, but I did make a change based off of it.

The multiple lines are now replaced with `text = ' '.join([paragraph['text'] for paragraph in data['body_text']])`